api-call.py

```diff
@@ -1,29 +1,322 @@
+                    'url_textbooks': self.config['apiUrls']['textbooks'],
+                    'parameters': {'date': user_date},
+                    'header': header}
+        elif choice == '4':
+            route_id = input('Enter Route Number: ')
+            url_obj = {'url_routes': self.config['apiUrls']['routes'],
+                    'url_arrivals': self.config['apiUrls']['arrivals'],
+                    'url_vehicles': self.config['apiUrls']['vehicles'],
+                    'route_id': route_id,  'parameters': {},
+                    'header': header}
+        else:
+            print('Invalid choice. Please try again.')
+
+        return url_obj
+
+    def get_text_books_with_term_date(self, url_obj: Dict[str, Any]) -> None:
+        """Get text books based on specified term and date
+        """
+
+        terms_data_response = self.get_api_data(
+            url_obj['url_terms'],
+            url_obj['parameters'],
+            url_obj['header']
+        )
+        if terms_data_response:
+            terms_data = terms_data_response.get('data', [])
+            if terms_data:
+                first_term = terms_data[0]
+                if 'attributes' in first_term:
+                    attributes = first_term['attributes']
+                    calendar_year = attributes.get('calendarYear')
+                    season = attributes.get('season')
+                    if calendar_year is None:
+                        print("Error: 'calendarYear' attribute is missing or empty")
+                    if season is None:
+                        print("Error: 'season' attribute is missing or empty")
+                else:
+                    print("Error: 'attributes' key is missing in the first term object")
+            else:
+                print("Error: 'data' key is missing or empty")
+
+    def get_stops_vehicles_on_route(self, url_obj: Dict[str, Any]) -> None:
+        """Get stops and vehicles on a given route
+        and determine where it is heading in real time
+        and the ETA for the stop
+        """
+
+        routes_data_response = self.get_api_data(
+            f"{url_obj['url_routes']}/{url_obj['route_id']}",
+            url_obj['parameters'],
+            url_obj['header']
+        )
+        if routes_data_response:
+            if 'data' in routes_data_response:
+                route_data = routes_data_response['data']
+
+                if 'attributes' in route_data:
+                    route_attributes = route_data['attributes']
+
+                    if 'description' in route_attributes:
+                        route_name = route_attributes['description']
+                    else:
+                        print("Error: 'description' key is missing in route attributes")
+                else:
+                    print("Error: 'attributes' key is missing in route data")
+            else:
+                print("Error: 'data' key is missing in routes data")
```

Comment:
These attributes checking logic are repeated. Can you have a separate function to do these check instead of wrapping your code with similar logic multiple times?

Reply:
handled with a utility function.